contracts/src/interfaces/IStakeTable.sol

```diff
@@ -0,0 +1,127 @@
+// SPDX-License-Identifier: Unlicensed
+
+pragma solidity ^0.8.0;
+
+import { BN254 } from "bn254/BN254.sol";
+import { EdOnBN254 } from "../libraries/EdOnBn254.sol";
+
+/// @title Interface for stake table that keep track of validators' external key and staked amount.
+/// @dev Stake delegation happens in a separate DelegationPool contract, and specific to
+/// instantiation thus not part of this interface
+interface IStakeTable {
+    /// @notice Supported stake type, either using native token or re-staked using ETH
+    enum StakeType {
+        Native,
+        Restake
+    }
+
+    /// @dev (sadly, Solidity doesn't support type alias on non-primitive types)
+    // We avoid declaring another struct even if the type info helps with readability,
+    // extra layer of struct introduces overhead and more gas cost.
+    //
+    /// @dev BLS verification key (used for Consensus voting in HotShot) is `BN254.G1Point`.
+    // `type BlsVerKey is BN254.G1Point;`
+    //
+    /// @dev Verification key of a Schnorr signature is just a `EdOnBN254.EdOnBN254Point`
+    // `type SchnorrVerKey is EdOnBN254.EdOnBN254Point;`
+
+    /// @notice Represents a HotShot validator node
+    /// In the dual-staking model, a HotShot validator could have multiple `Node` entries.
+    /// @param account The Ethereum account of the validator.
+    /// @param stakeType The type of token staked.
+    /// @param balance The amount of token staked.
+    /// @param registerEpoch The starting epoch for the validator.
+    /// @param exitEpoch The ending epoch for the validator.
+    struct Node {
+        address account;
+        StakeType stakeType;
+        uint64 balance;
+        uint64 registerEpoch;
+        uint64 exitEpoch;
+    }
+
+    // === Table State & Stats ===
+
+    /// @notice Get the commitment of the stake table used in current voting (i.e. snapshot at
+    /// the start of last epoch)
+    function votingStakeTableCommitment() external view returns (bytes32);
+    /// @notice Get the commitment of the stake table frozen for change (i.g. snapshot at the start
+    /// of the current epoch)
+    function frozenStakeTableCommitment() external view returns (bytes32);
+
+    /// @notice Get the total stake of the registered keys in the latest stake table (Head)
+    function totalStake() external view returns (uint256);
```

Comment:
Should we differentiate between native and restake?

Reply:
very interesting point! I see these options:
1. keep as it is, let instantiations specify its own conversion, the output is neither native nor restaked total stake, but an interpreted/exchange-rate-weighted total stake
2. returns `(uint256, uint256)`, first one for native, second one for restaked
3. returns `(StakeType[] memory, uint256[] memory)`

Option 3 is the most flexible and future-proof API, since it allows StakeType to be extended to more types (e.g. `EthRestaked, BtcRestaked, ...`, but Option 2 is most straightforward if the assumption that we only support ETH and native token is valid for the foreseeable future.

WDYT?